### ISSUE TITLE
New framework for managing filters in populse

### ISSUE THREAD

**Issue by sapetnioc** [populse/populse_db#16]
This issue would require a better description but, for now, I will just start by posting a file that had been discussed in a meeting.

```
#---------------------------------------------
# Selection usage for parameter "input_image"
# of process "my_module.my_process" within
# "mia.path_selection" organization
#---------------------------------------------
from populse_db import Database

db = Database('sqlite:////somewhere/ma_base')
selector_factory = db.find_selector_factory('mia.path_selection')

image_selector = selector_factory(process='my_module.my_process', 
                                  parameter='input_image')
with db as session:
    for row in session.execute(image_selector):
        print(row)

#---------------------------------------------
# mia/path_selection/__init__.py
#---------------------------------------------
from populse_db.path_selection import SelectorFactory

selector_factory = SelectorFactory()
selector_factory.add_process_module('process_selectors', '.*')


#---------------------------------------------
# mia/path_selector/process_selectors.py
#---------------------------------------------
from ..path_selection import selector_factory

@selector_factory.process('my_module.my_process')
class MyProcessPathSelector:
    @staticmethod
    def input_image(db, select):
        return select.select(Column('modality') == 'MRI',
                             Column('format') in ['NIFTI-1', 'DICOM'],
                             in_list('MRI', Column('modality'))) )

#---------------------------------------------
# populse_db/database.py
#---------------------------------------------
class Database:

    ...

    def find_selector_factory(self, factory_name):
        # Import factory module
        try:
            __import__(factory_name)
        except ImportError as exception:
            raise ImportError('Cannot import selector factory module "{0}": {1}'
                            .format(factory_name, exception))
        factory_module = sys.modules[factory_name]
        
        # Find the 'selector_factory_class' in the module
        selector_factory_class = getattr(factory_module, 'selector_factory_class', None)
        if selector_factory_class is None:
            raise ImportError('Selector factory module "%s" does not contain '
                            '"selector_factory_class"' % factory_name)
        
        return selector_factory_class(factory_name, self)

#---------------------------------------------
# populse_db/path_selection.py
#---------------------------------------------
import sys
import re

class SelectorFactory(object):
    def __init__(self, name, database):
        self.name = name
        self.database = database
        self.process_modules = []
        self.process_selectors = {}
    
    def add_process_module(self, module_name, regexp):
        self.process_modules.append((module_name, re.compile(regexp)))
        
    def load_process_module(self, process_name):
        module_to_load = None
        for module_name, regexp in self.process_modules:
            if regexp.match(process_name):
               module_to_load = module_name 
        if module_to_load is None:
            raise ValueError('In selector factory "{0}", no module found '
                             'for process "{1}"'.format(self.name, process_name))
        __import__(module_to_load)
        return module_to_load
    
    def process(self, process_name):
        def register_class_decorator(cls):
            self.process_selectors[process_name] = cls
            return cls
        
    def __call__(self, process_name, parameter):
        selector_class = self.process_selectors.get(process_name)
        if selector_class is None:
            module_loaded = load_process_module(process_name)
            selector_class = self.process_selectors.get(process_name)
            if selector_class is None:
                raise ValueError('No process selector found for process "{0}" '
                                 'in module "{1}"'.format(process_name,
                                                          module_loaded))
        selector = getattr(selector_class, parameter, None)
        if selector is None:
            raise ValueError('Process selector "{0}" has no selector for '
                             'parameter "{1}"'.format(selector_class,
                                                      parameter))
        select = self.database.table_classes['path'].select()
        return selector(select)

```

**Comment by ouvrierl** [comment]
I found some problems while starting to write the filters in MIA:

* It crashes if the tag has spaces in the name (it is not read as a column name)

* I also tried to write == null or != null, but I akways have no column named null (I also tried with "null", 'null', ...)

Am I doing something wrong?

**Comment by ouvrierl** [comment]
I am ok with square brackets [], but if we do that, won't it interfere with list values?

**Comment by sapetnioc** [comment]
You are right. I believe we can remove ambiguity but for users looking for
an expression it is better to visually distinct lists from field name.

We could use curly brackets such as {a field name}

**Comment by ouvrierl** [comment]
I have something strange,

Here is the first filter: (((**DirectionDiffusionNumber != null**)) AND (name IN ["data/raw_data/Guerbet-C6-2014-Rat-K39-Tube9-2014-02-11_08-01-22-11-G4_Guerbet_T1SE_800-RARE__pvm_-00-01-42.400.nii", "data/raw_data/Guerbet-C6-2014-Rat-K39-Tube9-2014-02-11_08-01-22-03-1-G2_Guerbet_ADC-DtiEpi__pvm_-00-03-07.200.nii", ..."]))

The second filter is (((**DirectionDiffusionNumber == null**)) AND (name IN ["data/raw_data/Guerbet-C6-2014-Rat-K39-Tube9-2014-02-11_08-01-22-11-G4_Guerbet_T1SE_800-RARE__pvm_-00-01-42.400.nii", "data/raw_data/Guerbet-C6-2014-Rat-K39-Tube9-2014-02-11_08-01-22-03-1-G2_Guerbet_ADC-DtiEpi__pvm_-00-03-07.200.nii", ...]))

With exactly the same filter (only the condition on null is modified), I have random results, is it possible?

It is either working, or null is recognized as a field (most of the time).

**Comment by sapetnioc** [comment]
You are right, result is different from one run to another. Well, I have no
clue to what is going on yet. I suspect either a bug in Lark or an
ambiguity in the grammar that gets resolved using an information that can
vary on each execution (like memory address). I do not know how to fix this
yet.

**Comment by sapetnioc** [comment]
Ok, it seems related to a bug in Lark (see [this issue](https://github.com/lark-parser/lark/issues/81)) that had been reproduced by the author 28 days ago but not fixed yet.

**Comment by sapetnioc** [comment]
I was able to deal with the bug in the grammar parsing. The pseudo-random trees we can get from grammar parsing all gives the same query. When converting a field name, I first check that it is not a literal (it should be done by grammar, but...).

**Comment by ouvrierl** [comment]
I was also working on it, but you have been faster than me :)

Do you plan on working on the operator (I)LIKE, or I can do it?

**Comment by sapetnioc** [comment]
Sorry, I did not know you was trying to fix this. You can do LIKE/ILIKE if you wish or I will do it. Tell me. By the way, I went back on a few renaming you made in `filter.py`. Some `column` really mean SqlAlchemy column (and not field) and some `.name` are SqlAlchemy attributes (not default primary key). I saw a few `tag` left. You can remove them or I will do it later.

**Comment by ouvrierl** [comment]
No problem, it's no big deal, and sorry for the renaming mistakes, I will try to improve it this afternoon.

And I will probably try the LIKE operator today, I will update you

**Comment by ouvrierl** [comment]
I tried to add operator.contains in python_operators for LIKE operator, but I get a sqlalchemy error saying that :

NotImplementedError: Operator 'contains' is not supported on this expression

Should we create our own operator? I must admit I am a little lost

I can't find what is precisely happening in sqlalchemy when .like is called on the column (a sql LIKE statement is generated, that's sure)

**Comment by sapetnioc** [comment]
For standard operators the same code works for both Python and SqlAlchemy
thank to operator overloading. For other opertors such as like or ilike
this is not possible. To go further, we should have two different
dictionaries, the existing one (python_operators) and one for SqlAlchemy
operators. Like and ilike would probably use custom function or lambda
function for Sql.

**Comment by ouvrierl** [comment]
I found a way to do it by using sqlalchemy like operators, you can look at master branch

With this working, it raised another problem for the rapid search if I have both simple and list types in the databrowser, as I have this error :

Combination of simple fields conditions with list fields conditions is only allowed with AND but not with or

Do you think it will be possible to perform this kind of filter? Because we need it in MIA

**Comment by sapetnioc** [comment]
To date the implementation of filter first uses SQL to select documents and then Python to do further selection that cannot be done with SQL. This is a kind of hardcoded AND between SQL and Python expression. Since the list operations cannot be implemented in SQL, the only possibility I see to make the request working with OR is to make it fully in Python. It may be slow because we will scan all documents from database and apply a Python filter to each one. 

We could make the process of using a full Python expression completely transparent to the user. Is it a good idea ? It would mean that if we make a mistake in the way we write a filter (for instance using a list condition before a simple field condition), we would get a slow Python function without knowing that it could be possible to make it faster by reorganizing filter conditions.

The answer could be to make it work, whatever it costs to performance, and switch to Postgres when we have performance problems (I think we will able to implement filters fully in SQL with Postgres).

**Comment by denisri** [comment]
By doing this we would completely lose the benefit of using a database. We know that python is not good at iterating over large collections, and we have the experience of requests that can last several minutes or more, freezing the software and ruining the user experience (and people's opinion about the software), this is one of the main reasons we wanted to move away from the former brainvisa implementation. I generally agree that we should first make things work before optimizing them, but, well, this is an exception because it can be really really harmful. So I would not like to allow such inefficient requests, or at least I would like to have a means of blocking them.

**Comment by sapetnioc** [comment]
We may have another option that could make all list selection feature work with SQLite without touching user API. Moreover, it should remove all the SQL + Python expression combination code. If we add a table for each list field containing the list values (we had such tables before), we could use them to implement efficient IN operator in SQL. This was not possible without a custom query language but thanks to this language, we have the possibility to translate the IN operations into subqueries. For instance a filter like `"value" IN list_field` could translate into something like `SELECT * FROM the_collection WHERE 'value' IN (SELECT value FROM _the_collection_list_field WHERE _the_collection_list_field.id = the_collection.id) `.

The idea is simple but I suppose the implementation may be not as simple as it seem. Especially because of the separation between the WHERE clause (built from the filter parsing) and the SELECT clause (added later when selecting the collection on to apply the filter). In the example, we must reference in the WHERE clause a field from the SELECT clause. I do not know yet a way to build the subquery of the filter in SqlAlchemy without accessing the table corresponding to `the_collection`. But the filter does not reference this collection, it could be applied to any collection. We may have to change the API a little bit by providing a collection name for parsing filters.

**Comment by ouvrierl** [comment]
Having a table for each list field has already been tested, and is not acceptable in execution time (I think I divided by 10 the execution time when removing those tables).

Otherwise, can't we split the filter in 2 subfilters in order to execute OR between simple fields and list fields? At the end we would get 2 lists of documents and we would just have to do an union between those 2 lists no?

**Comment by sapetnioc** [comment]
The idea would be to have both the complete list in the main table (as
today) and another copy of the list values in another table. This would not
affect performances for querying but insertion will be slower. Your are
right, we have to deal with performance issues. The good news is that
whatever the solution we choose, the API will not be changed. We could even
provide several implementations according to database parameters. For
instance letting the user favor insertion speed or querying speed.
Therefore, I propose to make OR filter work with a first full Python
implementation and then work on optimization eventually providing alternate
implementations.

Even if we split the query in two, the Python part will need to check all
documents against its filter. Therefore, this will be the speed bottleneck
anyway. So we will be more efficient in putting the whole filter in Python
and doing only one test per document.

**Comment by sapetnioc** [comment]
I made a script that generate data and added an option to add a table for each list field (also keeping the list as a string in the main table). The speed factor when we add list tables is roughly between 120% and 280%. It is clearly related to the cumulated size of lists in the whole database. For the worst case (280%) I generated one list field for each standard field. The list sizes are random between 0 and 100 and all items are randomly generated. Therefore, the database has roughly 50 x more data in lists than in standard fields. For 120%, I created 20% of list fields (2 list fields for 10 standard fields). This still represent a lot of data in lists.

To my opinion, we may give a try to this implementation. I did not implemented querying and did not commit anything, I just tried locally for testing insertion speed. I will not touch anything before the collections are done to avoid merging problems.

**Comment by ouvrierl** [comment]
Ok I will try to be fast then, I am almost done, currently writing some tests to make it more robust

**Comment by sapetnioc** [comment]
Good. By the way for my tests I added the possibility to give a document dictionary to add_document, making it possible to have only one query for the main table. An for lists, I also used a single `self.session.execute` per list field (giving all values of row to insert in a list). This is very little change to existing code.

**Comment by ouvrierl** [comment]
I also implemented the possibility to give a dict to add_document, we can either pass a dict of values, or a str corresponding to the primary_key value (the code is not pushed yet)

**Comment by ouvrierl** [comment]
I just pushed everything

**Comment by sapetnioc** [comment]
I pushed the implementation that adds the list tables. I activated it by default so that you can check if insertion time is acceptable for MIA. To make comparison with former implementation, it is possible to deactivate list tables at run time by using `list_tables=False` at database creation.

If this implementation is not too slow for users, I will make it mandatory and put all the filtering in SQL. That will make it possible to use any AND/OR combination in filters.

**Comment by ouvrierl** [comment]
Ok thanks!

I just merged everything with what I was doing (I just had a problem with circular imports so I had to import populse_db only)

I will do profiling and see if it's acceptable like this next week, but from what I saw rapidly, it takes more time

**Comment by ouvrierl** [comment]
With the tables for lists, the import now takes around 5 seconds for 200 scans in MIA, and with profiling, I saw that 70% of the execution time is to execute the request that insert the list values (self.session.execute(sql, params=sql_params))

**Comment by sapetnioc** [comment]
So it is more than two times slower. Do you have many list tags comparing to simple tags ? I suppose the insertion speed is a problem for MIA, no ?

The goal of all this is to find a solution to implement the OR on filters using lists tag. I see the following options :

1. Include list tables in data model as implemented today and use only SQL for filters.
1. Include list tables but create them only if they are used in a filter. The time spend for filling list tables would not be spent during insertion but it would be during execution of some filters.
1. Drop list tables and implement filters in SQL + Python with a possibility some filters fully in Python that can be very slow.
1. Implement both models (with and without list tables) and add the possibility to switch the model of an existing database. If a database without list tables is opened with `list_tables=True`, the list tables would be created (it may be slow) and the filter implementation would be SQL. On the other hand, if a database with list tables is opened with `list_tables=False`, the list tables would be dropped and the filters would be SQL + Python (or Python only depending on the filter). Not giving explicitly `list_tables` during database opening would use the current model of the database.

**Comment by ouvrierl** [comment]
After a long talk with Eric, we decided that it would be best to take the fourth solution for now.
It would give us the possibility to make real tests (we were thinking about around 1000 scans) and have numbers to compare solutions 1 and 3, for insertion and filters.
For the second solution, it would be the fourth solution with the flag list_tables=False, and a function before the filter that creates the missing list tables, in order to be able to do only SQL.

To put things in a nutshell, we would have to implement:
 * Filters without the list tables (Python if SQL not possible)
 * Filters with the list tables (SQL only)
 * Function that creates the missing tables at each filter (for solution 2)
 * Function managing the list_tables flag, that creates or removes all the list tables 

Am I missing something?

**Comment by sapetnioc** [comment]
There was a bug in list tables implementation. I was not iterating over list values but over each character of the whole list converted in a string. Therefore I inserted too many rows and I do not know how it did not fail. Sorry about that but since I had not checked querying until today, I did not see it before. So it may be faster now.

I am working on the filters in SQL only now (it will be needed anyway). The most difficult part for the solution 4 would be to implement the full Python switch in the case SQL is not possible. There is no need to create tables during filtering in solution 4 because the implementation will be chosen at connection time with `list_tables=True|False`. If the implementation is not the appropriate to the model, we have choices:
1. raise an exception. We would provide a method to explicitly change the model.
1. silently change the model (adding or removing the list tables)

**Comment by ouvrierl** [comment]
Thank you for fixing the list tables, I will take care of silently changing the model for now, we will see later about solution 2 that needs to add tables at filter

**Comment by sapetnioc** [comment]
Could you first check that there is no substantial speed-up with the fix ? Because if there is a chance to have only one filter implementation, we should catch it. Otherwise, I can make both filter implementations.

**Comment by sapetnioc** [comment]
A new problem. I do not know how to implement ordering comparison operators (<, >, <=, >=) on lists in SQL. I can compare the string representation of the list but it is not correct (because `[1]` < `[10]` but `"[1]"` > `"[10]"`). I could eventually compare the first N elements (N being fixed before executing the query). But I do not know how to compare elements of lists with various lengths.

I think I will change a little bit the filtering API to let the user choose the query creation given a filter. There would be an option to select the kind of request that can be build :

* `query_type='sql'`: build only a SQL query. If it is not possible, raises an error
* `query_type='python'`: build only a Python query. Will always work but will not use the database to select documents (could be slow)
* `query_type='mixed'`: build a SQL query with eventually a Python query for post-selection. If not possible, raises an error.
* `query_type='guess'`: build a mixed query. If not possible, build a Python query.

I propose to use `mixed` as default value for this parameter. The default value could also be changed for a whole database instance (e.g. a parameter in `__init__`).

**Comment by ouvrierl** [comment]
We decided to precisely quantify the execution time for the import in MIA, with and without list tables, in order to decide if it's acceptable or no. I will tell you more about this as soon as I have precise numbers.

Maybe we could leave the Python implementation of the filters for now, to see if we really need it or no.

For the list operators, we were thinking about comparing a list and a value, as comparing 2 lists with < and > operators does not have much meaning.
If there is FOV (float list) > 2 for example, it would give the list of documents having at least 1 FOV value > 2.

And I am ok for the query_type flag, it would leave the choice (at least for now)

**Comment by sapetnioc** [comment]
I think I just finished full Python implementation (not commited yet). We
will keep it anyway if ever we need a pure Python filter. I reorganized the
code to make it rather simple to specialize implementation without recoding
the whole filter parsing.

**Comment by ouvrierl** [comment]
Ok so here is the precise execution time for the import in MIA

Without tables:
 * 1.95 sec for 1 3D scan
 * 2.8 sec for 200 3D scans
 * 6.65 sec for 1000 3D scans
  * Which give us the affine function 1.95 sec + 0.9 sec/200 3D scans


With tables:
 * 2.68 sec for 1 3D scan
 * 4.78 sec for 200 3D scans
 * 14.3 sec for 1000 3D scans
  * Which give us the affine function 2.68 sec + 2.2 sec/200 3D scans

**Comment by servoz** [comment]
@ouvrierl , thanks for this very precise and useful results.
it turn out that we have to decide if we accept these orders of time ? For local use it should be ok (let says 1min for 4000 scans, with table). But for big data ? May be a good point of view could be to propose a light tool for local use, using SQL (MIA), on limited data and other tool (that could be also MIA) using PostgreSQL dedicated to big data .. 
in 16#issuecomment-394302051

**Comment by denisri** [comment]
It's not too bad actually... For big data I guess insertion will be done
only once, but many queries can be made afterwards. 1 min for 4000 scans is
acceptable I think - actually more time will be spent on other tasks during
importation (conversion, data checking...) and database insertion will not
be the bottleneck.


2018-06-05 19:10 GMT+02:00 Eric Condamine <notifications@github.com>:

**Comment by sapetnioc** [comment]
The good news is that I have both the Sql, Python and Mixed implementations. The Guess one will be just a matter of try/except. The bad news is that I found a fundamental difference in the way Python and SQLite behave when selecting empty fields (`None` for Python and `NULL` for SQL). Let's take an example, if I store the 4 following documents with two fields `x` and `y`:
- `x=0 y=None`
- `x=1 y='A'`
- `x=2 y='DICOM'`
- `x=3 y='NIFTI`

We expect that the filter `y < 'DICOM'` returns documents 0 and 1 and that `y >= 'DICOM'`returns 2 and 3. No problem with Python for that. But in SQL the condition simply ignore fields that are `NULL`. Therefore, `y < 'DICOM'` only returns 1 and  `y >= 'DICOM'`returns 2 and 3. According to the following test, it is a feature of SQLite :
```
sqlite> create table toto (x, y);
sqlite> insert into toto values (0, NULL);
sqlite> insert into toto values (1, 'A');
sqlite> insert into toto values (2, 'DICOM');
sqlite> insert into toto values (3, 'NIFTI');
sqlite> select * from toto where y < 'DICOM';
1|A
sqlite> select * from toto where y >= 'DICOM';
2|DICOM
3|NIFTI
sqlite> select * from toto;
0|
1|A
2|DICOM
3|NIFTI
```
The behavior is exactly the same for PostgreSQL. What should we do ? Port the SQL behavior to Python (easy) or port the Python behavior to SQL (much more complex) ? I have a small preference for the easy solution even if I do not like this behavior.

**Comment by denisri** [comment]
I suppose SQLAlchemy behaves like SQL since it is actually SQL under the hood. And nobody has ever complained.
I think we should behave like SQL also, as long as it is documented. We are doing databasing here, so let's behave like a database, with its fastest implementation by default. If we want to query null values in addition to "< something" we should do a more complex query with a OR.

**Comment by sapetnioc** [comment]
I comited the new filtering API with four query types : sql, python, mixed and guess. The Python behavior is aligned to the SQL behavior (we can still discuss that point though).

**Comment by sapetnioc** [comment]
My next move will be to run the tests several times with various database parameters (list tables or not, query type = mixed, python or guess). I have some code somewhere to deal with reusing the same test class several times with various parameters.

**Comment by ouvrierl** [comment]
I also think it's best to keep the SQL behavior, and we can add conditions to the filter if we want to check NULL values with "null" right?

As for the tests, it would be indeed great to automatically test all parameters.
I already found a test failing without the caches (test_remove_field), I am trying to fix it right now.

**Comment by ouvrierl** [comment]
We established a list of situations that do not have meaning, and we would like your opinion:
 * >, <, >=, <=, (and between) for boolean, string, and all lists
 * IN (in a set of values) for lists

Do you agree that those searches do not have meaning? Do you see other situations?

In MIA, we will forbid those searches, should we forbid it in MIA or directly in populse_db filtering?

**Comment by ouvrierl** [comment]
I found out some interesting behaviors of the filter, and wanted to know if it's normal
 * with query_type="python", it crashes with the operator like, because it's not in self.python_operators
 * with all query_types, it always returns an empty generator if in operator is used

**Comment by sapetnioc** [comment]
We established a list of situations that do not have meaning, and we would
like your opinion:

   -

   , <, >=, <=, (and between) for boolean, string, and all lists


It can be meaningful for strings. I would add that these comparison between
two different types (e.g. int and str) is not meaningful.


   - IN (in a set of values) for lists

Do you agree that those searches do not have meaning? Do you see other
situations?

In MIA, we will forbid those searches, should we forbid it in MIA or
directly in populse_db filtering?

Some are already forbidden by filtering because the structure is detected
during parsing (e.g. operator in between lists). For other cases, I would
prefer not to alwayd include testing validity during filter parsing because
it can takes time when using a lot of existing valid filters. But we can
add a validate_filter() method that would only do the validation.

**Comment by ouvrierl** [comment]
For now, I still have the issue with IN operator, it is strange because the filters are like the tests.

We use IN operator in the advanced search in order to filter the list of scans to search into.
Here is the filter:
((({BandWidth} == "25000")) AND ({FileName} IN ["data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-02-G1_Guerbet_Anat-RARE__pvm_-00-02-20.000.nii", "data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-04-G3_Guerbet_MDEFT-MDEFT__pvm_-00-09-40.800.nii", "data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-05-G4_Guerbet_T1SE_800-RARE__pvm_-00-01-42.400.nii", "data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-06-G4_Guerbet_T1SE_800-RARE__pvm_-00-01-42.400.nii", "data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-08-G4_Guerbet_T1SE_800-RARE__pvm_-00-01-42.400.nii", "data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-09-G4_Guerbet_T1SE_800-RARE__pvm_-00-01-42.400.nii", "data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-10-G3_Guerbet_MDEFT-MDEFT__pvm_-00-09-40.800.nii", "data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-11-G4_Guerbet_T1SE_800-RARE__pvm_-00-01-42.400.nii"]))

It is supposed to return data/raw_data/Guerbet-C6-2014-Rat-K52-Tube27-2014-02-14_10-23-17-04-G3_Guerbet_MDEFT-MDEFT__pvm_-00-09-40.800.nii, but it returns no scan.

Even if I remove the scan filtering, and just do this query, I get no scan but I should have the same scan as in the previous query:
((({BandWidth} IN ["25000"])))

Do you know where the issue might be? Am I giving good filters or the parsing is wrong?

**Comment by sapetnioc** [comment]
Your filter seems good. I suspected a literal parsing issue but I copy-pasted your file name list in literal tests and it is ok. Therefore it is necessary to go further in debug. Could you provide a code sample that show the bug ?

**Comment by sapetnioc** [comment]
What about closing this issue. I think we reached a functional state for filter so we can start creating different issues for various questions and bugs. I would add two new issues : one to decide what to do with table lists (make mandatory, remove or make optional) and one for the filtering bug.

I will comment other issues that may be closed too.

**Comment by ouvrierl** [comment]
No problem, I will write a new issue for the bug with IN operator
